flow/hstore/hstore.go

```diff
@@ -0,0 +1,240 @@
+/*
+This is in reference to PostgreSQL's hstore:
+https://github.com/postgres/postgres/blob/bea18b1c949145ba2ca79d4765dba3cc9494a480/contrib/hstore/hstore_io.c
+
+This package is an implementation based on the above code.
+*/
+package hstore_util
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+type text struct {
+	String string
+	Valid  bool
+}
+
+type hstore map[string]*string
+
+type hstoreParser struct {
+	str           string
+	pos           int
+	nextBackslash int
+}
+
+func newHSP(in string) *hstoreParser {
+	return &hstoreParser{
+		pos:           0,
+		str:           in,
+		nextBackslash: strings.IndexByte(in, '\\'),
+	}
+}
+
+func (p *hstoreParser) atEnd() bool {
+	return p.pos >= len(p.str)
+}
+
+// consume returns the next byte of the string, or end if the string is done.
+func (p *hstoreParser) consume() (b byte, end bool) {
+	if p.pos >= len(p.str) {
+		return 0, true
+	}
+	b = p.str[p.pos]
+	p.pos++
+	return b, false
+}
+
+func unexpectedByteErr(actualB byte, expectedB byte) error {
+	return fmt.Errorf("expected '%c' ('%#v'); found '%c' ('%#v')", expectedB, expectedB, actualB, actualB)
+}
+
+// consumeExpectedByte consumes expectedB from the string, or returns an error.
+func (p *hstoreParser) consumeExpectedByte(expectedB byte) error {
+	nextB, end := p.consume()
+	if end {
+		return fmt.Errorf("expected '%c' ('%#v'); found end", expectedB, expectedB)
+	}
+	if nextB != expectedB {
+		return unexpectedByteErr(nextB, expectedB)
+	}
+	return nil
+}
+
+func (p *hstoreParser) consumeExpected2(one byte, two byte) error {
+	if p.pos+2 > len(p.str) {
+		return errors.New("unexpected end of string")
+	}
+	if p.str[p.pos] != one {
+		return unexpectedByteErr(p.str[p.pos], one)
+	}
+	if p.str[p.pos+1] != two {
+		return unexpectedByteErr(p.str[p.pos+1], two)
+	}
+	p.pos += 2
+	return nil
+}
+
+var errEOSInQuoted = errors.New(`found end before closing double-quote ('"')`)
+
+// consumeDoubleQuoted consumes a double-quoted string from p. The double quote must have been
+// parsed already. This copies the string from the backing string so it can be garbage collected.
```

Comment:
I don't think it's important to copy the string like that; the map we construct doesn't escape the function after being encoded to json, so we aren't creating a long lived reference